keps/sig-testing/3041-node-conformance-and-features/README.md

```diff
@@ -0,0 +1,759 @@
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The document started as an analysis of whether `NodeFeature` label needs to be
+renamed to simply `Feature` as it carries an identical semantics [kubernetes/kubernetes#94289](https://github.com/kubernetes/kubernetes/issues/94289),
+but looking into it, scope needed to be extended into other labels currently
+being used, see also [this discussion](https://groups.google.com/g/kubernetes-sig-testing/c/8NVCt9UoV_Q/m/Vqz3_TidAwAJ?utm_medium=email&utm_source=footer).
+
+Document proposes changes into `Feature` label description, and additional labels
+that will make it clearer for contributors how to apply these labels.
+
+## Motivation
+
+Tests have various clarifiers - they may be testing the feature that is currently
+in development and depends on feature gate, test may require a special environment,
+or test may only work with specific hardware. Today a few labels are used to represent
+the combination of these "dimensions", in most cases a label `[Feature:]` is used.
+
+The universal nature of a single `[Feature:]` label makes it hard to apply it
+consistently and query tests.
+
+There are a few specific problems we saw recently:
+
+- There is no way to disable tests for a specific feature gate
+  (example where it’s needed: https://github.com/kubernetes/kubernetes/issues/99854).
+  It is a "feature" guarded by feature gate, but not marked as "feature".
+- There are no tests validating that k8s works with ALL beta feature gates disabled.
+  Creating such test would be hard with today's test labeling.
+- Some tests are degraded in how `[Feature]` and `[NodeFeature]` labels are applied
+  (for example: https://github.com/kubernetes/kubernetes/pull/105921).
+- labels that we apply heavily depend on the environment tests are running on.
+  For example, https://github.com/kubernetes/kubernetes/pull/104803 is not marked
+  as `Feature`, even though it depends on the environment - if `test-handler`
+  is not installed, test will not succeed. Similar thing to AppArmor tests.
+
+### Goals
+
+Split the special environment and feature development labels:
+
+- Clarify the semantic of a `[Feature:]` label.
+- Introduce the `[FeatureGate:]` label with the stage (`[Alpha]`, `[Beta]`, etc.)
+  addition, replace tags like `[NodeAlphaFeature]`.
```

Comment:
We should soon be based on Ginkgo v2 in Kubernetes 1.25 (see https://github.com/kubernetes/kubernetes/pull/109111).

Would it make sense to take advantage of that and move away from embedding tags inside the description text and instead use [Label decorators](https://pkg.go.dev/github.com/onsi/ginkgo/v2#Label)?

Reply:
I thought the same. We still will need to have well-known labels documented and provide some guidance around them

Reply:
So should clarifying labels and the transition to label decorators be separate KEPs or can we do it in one?

Reply:
I think we should transition to label decorators separately, clarifying these is valuable as-is and shouldn't have to block on the large ginkgo v2 PR and sorting out the migration details.

Reply:
added non-goal